File: backend/app/services/telegram_alerts.py
"""
Telegram trigger alerts: reads TELEGRAM_BOT_TOKEN + TELEGRAM_CHAT_ID from env.
Edge detection on the script-level last-action trigger band (row_id 0 in state map).
"""
from __future__ import annotations

import logging
import os
from threading import Lock
from decimal import Decimal

import httpx
from sqlalchemy.orm import Session

from app.models.execute_row_model import ExecuteRow
from app.models.script_model import Script

logger = logging.getLogger(__name__)

# (script_id, row_id) -> last buy/sell zone flags. row_id 0 = script-level last-action / flat band.
_edge_state_row: dict[tuple[int, int], dict[str, bool]] = {}
_visual_memory: dict[int, dict[str, bool]] = {}
_edge_state_lock = Lock()


def _telegram_configured() -> bool:
    token = (os.getenv("TELEGRAM_BOT_TOKEN") or "").strip()
    chat = (os.getenv("TELEGRAM_CHAT_ID") or "").strip()
    return bool(token and chat)


def send_telegram_message(text: str) -> bool:
    """POST sendMessage. Returns True on success. Never raises."""
    token = (os.getenv("TELEGRAM_BOT_TOKEN") or "").strip()
    chat_id = (os.getenv("TELEGRAM_CHAT_ID") or "").strip()
    if not token or not chat_id:
        return False
    url = f"https://api.telegram.org/bot{token}/sendMessage"
    try:
        with httpx.Client(timeout=12.0) as client:
            r = client.post(url, json={"chat_id": chat_id, "text": text[:4000]})
            if r.status_code >= 400:
                logger.warning("Telegram sendMessage failed: %s %s", r.status_code, r.text[:200])
                return False
            return True
    except Exception as e:
        logger.warning("Telegram request error: %s", e)
        return False


def clear_row_edge_state(script_id: int, row_id: int) -> None:
    """Remove edge memory for one row id (legacy); prefer clear_all_edge_state_for_script."""
    with _edge_state_lock:
        _edge_state_row.pop((int(script_id), int(row_id)), None)


def clear_all_edge_state_for_script(script_id: int) -> None:
    """Reset only edge-state latches for a script (re-arm detection)."""
    sid = int(script_id)
    with _edge_state_lock:
        for k in list(_edge_state_row.keys()):
            if k[0] == sid:
                del _edge_state_row[k]


def get_trigger_visual_states(
    script_id: int,
    current: Decimal | None,
    buy_t: Decimal | None,
    sell_t: Decimal | None,
) -> dict[str, str]:
    """Return UI states: NORMAL | ACTIVE | TRIGGERED for buy/sell."""
    sid = int(script_id)
    with _edge_state_lock:
        vm = _visual_memory.setdefault(sid, {"buy_hit_once": False, "sell_hit_once": False})
        buy_inside = bool(current is not None and buy_t is not None and current <= buy_t)
        sell_inside = bool(current is not None and sell_t is not None and current >= sell_t)
        if buy_inside:
            buy_state = "ACTIVE"
        elif vm["buy_hit_once"]:
            buy_state = "TRIGGERED"
        else:
            buy_state = "NORMAL"
        if sell_inside:
            sell_state = "ACTIVE"
        elif vm["sell_hit_once"]:
            sell_state = "TRIGGERED"
        else:
            sell_state = "NORMAL"
    return {"buy_state": buy_state, "sell_state": sell_state}


def emit_fresh_session_if_applicable(
    script: Script,
    ist_calendar_advanced: bool,
    rows: list,
    current: Decimal | None,
    buy_t: Decimal | None,
    sell_t: Decimal | None,
    pairs: list[tuple[Decimal, Decimal]],
) -> None:
    """
    One Telegram per script on the first IST day roll when there are no open holdings:
    sends market-based fresh buy/sell trigger band (same as dashboard aggregate).
    """
    if not _telegram_configured() or not ist_calendar_advanced:
        return

    from app.services.trigger_lifecycle import has_active_holdings

    if has_active_holdings(rows):
        return
    if current is None or buy_t is None or sell_t is None or not pairs:
        return

    send_telegram_message(
        f"🌅 Fresh session (IST)\n"
        f"No open holdings from the prior day.\n"
        f"Script: {script.name}\n"
        f"LTP (reference): {current}\n"
        f"Buy trigger: {buy_t}\n"
        f"Sell trigger: {sell_t}"
    )


def _process_row_edge(
    script: Script,
    row_id: int,
    current: Decimal,
    buy_t: Decimal,
    sell_t: Decimal,
) -> None:
    """One edge step for a logical row (real execute row id, or 0 = aggregate band)."""
    sid = int(script.id)
    key = (sid, int(row_id))
    buy_z = current <= buy_t
    sell_z = current >= sell_t
    send_buy = False
    send_sell = False
    with _edge_state_lock:
        prev = _edge_state_row.setdefault(key, {"buy": False, "sell": False})
        vm = _visual_memory.setdefault(sid, {"buy_hit_once": False, "sell_hit_once": False})
        # Emit at most one side per evaluation to avoid dual messages.
        if buy_z and not prev["buy"]:
            send_buy = True
            vm["buy_hit_once"] = True
        elif sell_z and not prev["sell"]:
            send_sell = True
            vm["sell_hit_once"] = True
        prev["buy"] = buy_z
        prev["sell"] = sell_z

    label = f"Row #{row_id}" if row_id else "Last-action band"
    if send_buy:
        send_telegram_message(
            f"🔔 {label} — BUY side\n"
            f"Script: {script.name}\n"
            f"LTP: {current}\n"
            f"Buy trigger (≤): {buy_t}  |  Sell trigger (≥): {sell_t}"
        )
    if send_sell:
        send_telegram_message(
            f"🔔 {label} — SELL side\n"
            f"Script: {script.name}\n"
            f"LTP: {current}\n"
            f"Buy trigger (≤): {buy_t}  |  Sell trigger (≥): {sell_t}"
        )


def notify_row_after_save(db: Session, script: Script, row: ExecuteRow) -> None:
    """
    After creating/updating a row: if LTP already hits the current last-action band, send once
    and sync edge state so polling does not duplicate.
    """
    if not _telegram_configured():
        return

    from app.services import trigger_lifecycle as tl
    from app.services.market_service import fetch_stock_quote_nse

    all_rows = db.query(ExecuteRow).filter(ExecuteRow.script_id == script.id).order_by(ExecuteRow.id.asc()).all()
    ref = tl.last_action_ref_price(all_rows)
    if ref is None or ref <= 0:
        return
    price_f = fetch_stock_quote_nse(script.name)
    if price_f is None:
        return
    current = (Decimal(str(price_f))).quantize(Decimal("0.01"))
    up_pct = (Decimal(str(script.up_percent))).quantize(Decimal("0.01"))
    down_pct = (Decimal(str(script.down_percent))).quantize(Decimal("0.01"))
    bt, st = tl.triggers_from_reference(ref, up_pct, down_pct)
    _process_row_edge(script, 0, current, bt, st)

File: backend/supabase/functions/_shared/telegramAlerts.ts
import { sql } from "./db.ts";
import { Decimal, d } from "./decimal.ts";
import { fetchRowsForScript, type ExecuteRow } from "./executeRows.ts";
import type { Script } from "./scriptService.ts";

function telegramConfigured(): boolean {
  const token = (Deno.env.get("TELEGRAM_BOT_TOKEN") ?? "").trim();
  const chat = (Deno.env.get("TELEGRAM_CHAT_ID") ?? "").trim();
  return Boolean(token && chat);
}

/** POST sendMessage. Returns true on success. Never throws. */
export async function sendTelegramMessage(text: string): Promise<boolean> {
  const token = (Deno.env.get("TELEGRAM_BOT_TOKEN") ?? "").trim();
  const chatId = (Deno.env.get("TELEGRAM_CHAT_ID") ?? "").trim();
  if (!token || !chatId) return false;
  const url = `https://api.telegram.org/bot${token}/sendMessage`;
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ chat_id: chatId, text: text.slice(0, 4000) }),
      signal: AbortSignal.timeout(12000),
    });
    if (!res.ok) {
      console.warn("Telegram sendMessage failed:", res.status, (await res.text()).slice(0, 200));
      return false;
    }
    return true;
  } catch (e) {
    console.warn("Telegram request error:", e);
    return false;
  }
}

/** Reset only edge-state latches for a script (re-arm detection). */
export async function clearAllEdgeStateForScript(scriptId: number): Promise<void> {
  await sql`delete from trigger_edge_state where script_id = ${scriptId}`;
}

export type VisualStates = { buy_state: string; sell_state: string };

export async function getTriggerVisualStates(
  scriptId: number,
  current: Decimal | null,
  buyT: Decimal | null,
  sellT: Decimal | null,
): Promise<VisualStates> {
  const [vm] = await sql<{ buy_hit_once: boolean; sell_hit_once: boolean }[]>`
    insert into trigger_visual_memory (script_id, buy_hit_once, sell_hit_once)
    values (${scriptId}, false, false)
    on conflict (script_id) do update set script_id = excluded.script_id
    returning buy_hit_once, sell_hit_once
  `;

  const buyInside = current !== null && buyT !== null && current.lte(buyT);
  const sellInside = current !== null && sellT !== null && current.gte(sellT);

  const buyState = buyInside ? "ACTIVE" : vm.buy_hit_once ? "TRIGGERED" : "NORMAL";
  const sellState = sellInside ? "ACTIVE" : vm.sell_hit_once ? "TRIGGERED" : "NORMAL";
  return { buy_state: buyState, sell_state: sellState };
}

export async function emitFreshSessionIfApplicable(
  script: Script,
  istCalendarAdvanced: boolean,
  rows: ExecuteRow[],
  current: Decimal | null,
  buyT: Decimal | null,
  sellT: Decimal | null,
  pairs: [Decimal, Decimal][],
): Promise<void> {
  if (!telegramConfigured() || !istCalendarAdvanced) return;

  const { hasActiveHoldings } = await import("./triggerLifecycle.ts");
  if (hasActiveHoldings(rows)) return;
  if (current === null || buyT === null || sellT === null || pairs.length === 0) return;

  await sendTelegramMessage(
    `🌅 Fresh session (IST)\n` +
      `No open holdings from the prior day.\n` +
      `Script: ${script.name}\n` +
      `LTP (reference): ${current}\n` +
      `Buy trigger: ${buyT}\n` +
      `Sell trigger: ${sellT}`,
  );
}

/** One edge step for a logical row (real execute row id, or 0 = aggregate band). */
async function processRowEdge(
  script: Script,
  rowId: number,
  current: Decimal,
  buyT: Decimal,
  sellT: Decimal,
): Promise<void> {
  const sid = script.id;
  const buyZ = current.lte(buyT);
  const sellZ = current.gte(sellT);
  let sendBuy = false;
  let sendSell = false;

  await sql.begin(async (tx) => {
    const [prev] = await tx<{ buy_zone: boolean; sell_zone: boolean }[]>`
      insert into trigger_edge_state (script_id, row_id, buy_zone, sell_zone)
      values (${sid}, ${rowId}, false, false)
      on conflict (script_id, row_id) do update set script_id = excluded.script_id
      returning buy_zone, sell_zone
    `;
    // Emit at most one side per evaluation to avoid dual messages.
    if (buyZ && !prev.buy_zone) {
      sendBuy = true;
      await tx`
        insert into trigger_visual_memory (script_id, buy_hit_once, sell_hit_once)
        values (${sid}, true, false)
        on conflict (script_id) do update set buy_hit_once = true
      `;
    } else if (sellZ && !prev.sell_zone) {
      sendSell = true;
      await tx`
        insert into trigger_visual_memory (script_id, buy_hit_once, sell_hit_once)
        values (${sid}, false, true)
        on conflict (script_id) do update set sell_hit_once = true
      `;
    }
    await tx`
      update trigger_edge_state set buy_zone = ${buyZ}, sell_zone = ${sellZ}
      where script_id = ${sid} and row_id = ${rowId}
    `;
  });

  const label = rowId ? `Row #${rowId}` : "Last-action band";
  if (sendBuy) {
    await sendTelegramMessage(
      `🔔 ${label} — BUY side\n` +
        `Script: ${script.name}\n` +
        `LTP: ${current}\n` +
        `Buy trigger (≤): ${buyT}  |  Sell trigger (≥): ${sellT}`,
    );
  }
  if (sendSell) {
    await sendTelegramMessage(
      `🔔 ${label} — SELL side\n` +
        `Script: ${script.name}\n` +
        `LTP: ${current}\n` +
        `Buy trigger (≤): ${buyT}  |  Sell trigger (≥): ${sellT}`,
    );
  }
}

/**
 * After creating/updating a row: if LTP already hits the current last-action band, send once
 * and sync edge state so polling does not duplicate.
 */
export async function notifyRowAfterSave(script: Script, row: ExecuteRow): Promise<void> {
  if (!telegramConfigured()) return;

  const tl = await import("./triggerLifecycle.ts");
  const { fetchStockQuoteNse } = await import("./marketService.ts");

  const allRows = await fetchRowsForScript(script.id, "asc");
  const ref = tl.lastActionRefPrice(allRows);
  if (ref === null || ref.lte(0)) return;
  const priceF = await fetchStockQuoteNse(script.name);
  if (priceF === null) return;
  const current = d(priceF, 2);
  const upPct = d(script.up_percent, 2);
  const downPct = d(script.down_percent, 2);
  const [bt, st] = tl.triggersFromReference(ref, upPct, downPct);
  await processRowEdge(script, 0, current, bt, st);
}
